### ISSUE TITLE
die() and encodings on stderr => segfault

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#9058]
Migrated from [rt.perl.org#46327](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=46327) (status was 'resolved')

Searchable as RT46327$

**Comment by p5pRT** [comment]
### From smurf@noris.de

#### Created by smurf@noris.de



\.\.\. or\, how to segfault perl in two lines of code&#8203;:

\#\!/usr/bin/perl

\#use utf8; \# does not fix the problem
binmode\(STDERR\,"&#8203;:encoding\(latin1\)"\); \# required for segfault
binmode\(STDOUT\,"&#8203;:encoding\(latin1\)"\); \# not required for fault

print "\\374";  \# works; that's a German u\-umlaut
die "\\374";    \# segfaults

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl v5.8.8:

Configured by Debian Project at Wed Dec  6 23:17:41 UTC 2006.

Summary of my perl5 (revision 5 version 8 subversion 8) configuration:
  Platform:
    osname=linux, osvers=2.6.18.3, archname=i486-linux-gnu-thread-multi
    uname='linux saens 2.6.18.3 #1 smp sat nov 25 13:39:52 est 2006 i686 gnulinux '
    config_args='-Dusethreads -Duselargefiles -Dccflags=-DDEBIAN -Dcccdlflags=-fPIC -Darchname=i486-linux-gnu -Dprefix=/usr -Dprivlib=/usr/share/perl/5.8 -Darchlib=/usr/lib/perl/5.8 -Dvendorprefix=/usr -Dvendorlib=/usr/share/perl5 -Dvendorarch=/usr/lib/perl5 -Dsiteprefix=/usr/local -Dsitelib=/usr/local/share/perl/5.8.8 -Dsitearch=/usr/local/lib/perl/5.8.8 -Dman1dir=/usr/share/man/man1 -Dman3dir=/usr/share/man/man3 -Dsiteman1dir=/usr/local/man/man1 -Dsiteman3dir=/usr/local/man/man3 -Dman1ext=1 -Dman3ext=3perl -Dpager=/usr/bin/sensible-pager -Uafs -Ud_csh -Uusesfio -Uusenm -Duseshrplib -Dlibperl=libperl.so.5.8.8 -Dd_dosuid -des'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=define use5005threads=undef useithreads=define usemultiplicity=define
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -DTHREADS_HAVE_PIDS -DDEBIAN -fno-strict-aliasing -pipe -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -DTHREADS_HAVE_PIDS -DDEBIAN -fno-strict-aliasing -pipe -I/usr/local/include'
    ccversion='', gccversion='4.1.2 20061115 (prerelease) (Debian 4.1.1-20)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lgdbm -lgdbm_compat -ldb -ldl -lm -lpthread -lc -lcrypt
    perllibs=-ldl -lm -lpthread -lc -lcrypt
    libc=/lib/libc-2.3.6.so, so=so, useshrplib=true, libperl=libperl.so.5.8.8
    gnulibc_version='2.3.6'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.8.8:
    /etc/perl
    /usr/local/lib/perl/5.8.8
    /usr/local/share/perl/5.8.8
    /usr/lib/perl5
    /usr/share/perl5
    /usr/lib/perl/5.8
    /usr/share/perl/5.8
    /usr/local/lib/site_perl
    .


Environment for perl v5.8.8:
    HOME=/home/smurf
    LANG=en_US.UTF-8
    LANGUAGE (unset)
    LC_ALL=en_US.UTF-8
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/smurf/bin:/usr/local/bin:/usr/bin:/bin:/usr/bin/X11:/usr/games:/usr/pop/bin:/pop/technik/bin:/usr/X11R6/bin
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @eserte

"Matthias Urlichs" \(via RT\) \<perlbug\-followup@&#8203;perl\.org> writes&#8203;:

> \# New Ticket Created by  "Matthias Urlichs" 
> \# Please include the string&#8203;:  \[perl \#46327\]
> \# in the subject line of all future correspondence about this issue\. 
> \# \<URL&#8203;: http&#8203;://rt\.perl\.org/rt3/Ticket/Display\.html?id=46327 >
> 
> 
> 
> This is a bug report for perl from smurf@&#8203;noris\.de\,
> generated with the help of perlbug 1\.35 running under perl v5\.8\.8\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please enter your report here\]
> 
> \.\.\. or\, how to segfault perl in two lines of code&#8203;:
> 
> \#\!/usr/bin/perl
> 
> \#use utf8; \# does not fix the problem
> binmode\(STDERR\,"&#8203;:encoding\(latin1\)"\); \# required for segfault
> binmode\(STDOUT\,"&#8203;:encoding\(latin1\)"\); \# not required for fault
> 
> print "\\374";  \# works; that's a German u\-umlaut
> die "\\374";    \# segfaults
> 
> 
> \[Please do not change anything below this line\]
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-

If you need a workaround\, then try to upgrade the warn/die string to
have a utf8 flag&#8203;:

$ perl \-e 'binmode\(STDERR\,"&#8203;:encoding\(latin1\)"\);$x="\\374";utf8&#8203;::upgrade\($x\); die $x'
ü at \-e line 1\.

The segfault does not occur in perl5\.10\.0 anymore \(since patch
\#30213\)\, but the output looks somewhat ugly&#8203;:

$ perl5\.10\.0 \-e 'binmode\(STDERR\,"&#8203;:encoding\(latin1\)"\);$x="\\374";die $x'
\\x\{2087482d\}e line 1\.

For a real fix it seems to me that the Perl interna need a SV version
of Perl\_write\_to\_stderr\(\) which makes sure that
PERL\_WRITE\_MSG\_TO\_CONSOLE gets a utf8\-encoded character string\.

Regards\,
&nbsp;        Slaven

\-\- 
Slaven Rezic \- slaven \<at> rezic \<dot> de

Tk\-AppMaster&#8203;: a perl/Tk module launcher designed for handhelds
&nbsp;    http&#8203;://tk\-appmaster\.sf\.net

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From smurf@smurf.noris.de

On Wed Oct 10 15&#8203;:45&#8203;:51 2007\, slaven@&#8203;rezic\.de wrote&#8203;:

> If you need a workaround\, then try to upgrade the warn/die string to
> have a utf8 flag&#8203;:
> 
I found that already\, thanks\.

Unfortunately\, we're dealing with a whole lot of external input in a
legacy application which we're trying to make behave sanely in both UTF8
and latin1/latin9 environments; it'd be prohibitive to upfrade all the
inputs to utf8\. :\-\(

My current workaround is to call binmode\(STDERR\,"&#8203;:raw"\) and to just live
with the charset cruft\. Better than a segfault\. :\-/

> The segfault does not occur in perl5\.10\.0 anymore \(since patch
> \#30213\)\, but the output looks somewhat ugly&#8203;:
> 
> $ perl5\.10\.0 \-e 'binmode\(STDERR\,"&#8203;:encoding\(latin1\)"\);$x="\\374";die $x'
> \\x\{2087482d\}e line 1\.
> 
Ah yes\. I'm seeing this kind of error in 5\.8\.8 too\, when I replace the
latin1 encoding with a codec which just calls Encode\(\)\.

**Comment by p5pRT** [comment]
### From @cpansprout

This was fixed by 83c5555649\.

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
